### PR TITLE
[BUGFIX] Addresses issue #2993 (#3054) by using configuration when it is available instead of discovering keys (listing keys) in existing sources.

### PR DESCRIPTION
**Pull request by donaldheppner**
Changes proposed in this pull request:
- Use configuration to identify where key files should be in an existing store rather than querying the store for a list of potential keys and choosing those that match configuration.

Previous Design Review notes:
- Analysis of potential options can be found in issue comments: https://github.com/great-expectations/great_expectations/issues/3054

### Definition of Done
- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/data_context/data_context.py [hunk 1]
```diff
@@ -2432,13 +2439,17 @@ def validations_store_name(self):
...
-    def _compile_evaluation_parameter_dependencies(self):
+    def _compile_evaluation_parameter_dependencies(self, expectation_suite):
         self._evaluation_parameter_dependencies = {}
-        for key in self.expectations_store.list_keys():
-            expectation_suite = self.expectations_store.get(key)
-            if not expectation_suite:
-                continue
 
+        # only if we don't have an expectation suite do we do a key scan
+        if(expectation_suite is None):
```
Comment (talagluck) [inline great_expectations/data_context/data_context.py]: Why use this pattern instead of `if expectation_suite is None`?

File: great_expectations/data_context/data_context.py [hunk 2]
```diff
@@ -2432,13 +2439,17 @@ def validations_store_name(self):
...
+            for key in self.expectations_store.list_keys():
+                expectation_suite = self.expectations_store.get(key)
+                if not expectation_suite:
+                    continue
+
+        if(expectation_suite):
```
Comment (talagluck) [inline great_expectations/data_context/data_context.py]: Same here - why not `if expectation_suite`?

**Review by talagluck — COMMENTED** [review]
Hi @donaldheppner - thank you for this (and thank you for adding some much-needed type hinting as well)!

I just wanted to confirm - is this error only caused by the methods for storing EvaluationParameters? This is a less prominent store (as compared with Expectations, Validations, and DataDocs), and so that would interesting to know.

As far as testing, my inclination would be to add some tests to `tests/data_context/store/test_store_backends.py`. I believe this is where more generic StoreBackend tests would go, and so I believe it would be a good place to test the interaction of different StoreBackend configurations. Also, there is already some mock S3 testing infrastructure in place here! So hopefully that will make tests a bit simpler to implement.

Finally, it looks like the Azure pipeline might be failing due to linting issues. Would you mind running a linter and re-committing?

Thanks again for this! Looking forward to re-reviewing.

File: great_expectations/data_context/data_context.py (hunk 1 above)
Comment (donaldheppner) [reply great_expectations/data_context/data_context.py]: fixed

**Comment by donaldheppner** [comment]
I added the unit test (with more fixtures than I would have liked; thoughts on how I could either reuse validation operator fixtures or simplify the test setup are welcome); other issues raised have also been addressed.

**Review by talagluck — APPROVED** [review]
Thanks for the updates, @donaldheppner! This looks great. Agreed that the fixtures are a bit complex, but on a quick look-through, I didn't see anything current that was appropriate, so I don't want to gate on that for now.

This is awesome, and thank you for the thorough docstring as well. LGTM - congrats on your first PR into Great Expectations!

File: great_expectations/data_context/data_context.py [hunk 3]
```diff
@@ -2443,13 +2450,17 @@ def validations_store_name(self):
...
-    def _compile_evaluation_parameter_dependencies(self):
+    def _compile_evaluation_parameter_dependencies(self, expectation_suite):
         self._evaluation_parameter_dependencies = {}
-        for key in self.expectations_store.list_keys():
-            expectation_suite = self.expectations_store.get(key)
-            if not expectation_suite:
-                continue
 
+        # only if we don't have an expectation suite do we do a key scan
+        if expectation_suite is None:
```
Comment (OmriBromberg) [inline great_expectations/data_context/data_context.py:2457]: Hey this change actually breaks this [example](https://docs.greatexpectations.io/docs/guides/expectations/advanced/how_to_create_expectations_that_span_multiple_batches_using_evaluation_parameters#steps): 
after upgrading, I got the error of:
`Error while evaluating evaluation parameter expression: could not convert string to float: 'urn:great_expectations:validations:EXPECTATION_SUITE:EXPECTATION.result.observed_value:column=COLUMN_NAME'`
later on, I investigated and debugged GE to see what's going on, it turns out the Action '`StoreEvaluationParametersAction`' doesn't store previous metrics needed by future expectation suites since the `'_compile_evaluation_parameter_dependencies`' function is no longer calculating dependencies of all the expectation_suites but just the one running. please address this, not mentioned anywhere in release notes, willing to do the bugfix

Comment (OmriBromberg) [reply great_expectations/data_context/data_context.py:2457]: @donaldheppner @talagluck

**Review by OmriBromberg — COMMENTED** [review]
@donaldheppner @talagluck

File: great_expectations/data_context/data_context.py (hunk 3 above)
Comment (talagluck) [reply great_expectations/data_context/data_context.py:2457]: Hi @omribromberg - thanks so much for finding and flagging this. This was an unintended consequence. If you've already tracked this down and are up to making a fix, we would greatly appreciate it. I can prioritize reviewing this as soon as it's ready.

Comment (talagluck) [reply great_expectations/data_context/data_context.py:2457]: Also just a heads-up, I believe that #3660 which was recently filed is pointing to the same issue.

Comment (OmriBromberg) [reply great_expectations/data_context/data_context.py:2457]: @talagluck I think these might be different issues but might have the same solution, to solve evaluation parameters of this pattern: `urn:great_expectations:validations:{expectation_suite}:{expectation}.result.unexpected_percent:column={column_name}` which is what I was using, we'll need to go through all expectation suites' dependencies each time the action `StoreEvaluationParametersAction` is called to figure out if the current validation metrics needs to be saved for future use
correct me if I'm wrong, but removing the `expectation_suite` parameter from `_compile_evaluation_parameter_dependencies` would solve this with no apparent new bugs
lemme know if I can open a separate PR for this specific issue

Comment (talagluck) [reply great_expectations/data_context/data_context.py:2457]: Hi @OmriBromberg - thanks for the follow-up.

After some internal conversation, we believe that reverting this change is the right thing to do for the moment. It will possibly lead to another bug being resurfaced, but that bug is an edge case with a work-around (for which some more in-depth fixes are possible), whereas this one feels more urgent.

The fix you are proposing sounds right - it may also require the reversion of tests from this PR. Please do open a PR, and I will take a look once it's ready.

Thanks again!
